### PR TITLE
Added in Large-Batch SGD with a warmup, and a LARS startegy. Also add…

### PR DESCRIPTION
**Pull request by ashokei**
Large-Batch SGD with a warmup, and a LARS strategy. 

## Description ##
Added in Large-Batch SGD with a warmup, and a LARS strategy. Also added in a Polynomial Decay learning rate scheduler. Modified the example image fit code to allow these options to be selectable.

### Essentials ###
- [x ] Passed code style checking (`make lint`)
- [x ] Changes are complete (i.e. I finished coding on this PR)
- [x ] All changes have test coverage
- [ ] For user-facing API changes, API doc string has been updated. For new C++ functions in header files, their functionalities and arguments are well-documented. 
- [x ] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- Added in Large-Batch SGD with a warmup, and a LARS strategy. Also added in a Polynomial Decay learning rate scheduler. Modified the example image fit code to allow these options to be selectable.

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by piiswrong** [comment]
@zhreshold

**Comment by eric-haibin-lin** [comment]
@zhreshold could you help review?

File: python/mxnet/optimizer.py [hunk 1]
```diff
@@ -531,6 +531,197 @@ def update_multi_precision(self, index, weight, grad, state):
...
+            mult = maxmult
+        elif nwup <= 1:
+            mult = 1.0
+        else:
+            if (strategy == 'linear'):
+                mult = 1.0 + (maxmult - 1) * nup / nwup
+            elif (strategy == 'power2'):
+                mult = 1.0 + (maxmult - 1) * (nup * nup) / (nwup * nwup)
+            elif (strategy == 'power3'):
+                mult = 1.0 + (maxmult - 1) * (nup * nup) / (nwup * nwup)
```
Comment (zhreshold) [inline python/mxnet/optimizer.py]: Power3 is wrong

**Comment by zhreshold** [comment]
@ashokei See comments. Could you rebase and make pylint as well?

**Comment by ashokei** [comment]
@zhreshold thank you, will make requested changes.

**Comment by ashokei** [comment]
@zhreshold i made the requested changes, please review. thanks!

File: python/mxnet/lr_scheduler.py [hunk 2]
```diff
@@ -136,3 +136,42 @@ def __call__(self, num_update):
...
+    def __init__(self, num_update, max_update, base_lr=0.01, pwr=2):
+        super(PolyScheduler, self).__init__(base_lr)
+        assert isinstance(max_update, int)
+        if max_update < 1:
+            raise ValueError("maximum number of updates must be strictly positive")
+        self.base_lr_orig = self.base_lr
+        self.max_update = max_update
+        self.power = pwr
+        self.count = num_update
+        if num_update <= max_update:
```
Comment (zhreshold) [inline python/mxnet/lr_scheduler.py]: This this duplicate with line 173. I understand it is for resume training, but that should be handled in `__call__`, see the example in MultiFactorScheduler. Therefore, num_update is not necessary in `__init__`

Comment (ashokei) [reply python/mxnet/lr_scheduler.py]: @zhreshold i removed the num_update duplicate line, can you please check.thanks.

File: python/mxnet/lr_scheduler.py [hunk 3]
```diff
@@ -136,3 +136,38 @@ def __call__(self, num_update):
...
+    Parameters
+    ----------
+       num_update: current number of updates
+       max_update: maximum number of updates before the decay reaches 0.
+       base_lr:    base learning rate
+       pwr:   power of the decay term as a funtion of the current number of updates.
+
+    """
+
+    def __init__(self, num_update, max_update, base_lr=0.01, pwr=2):
```
Comment (zhreshold) [inline python/mxnet/lr_scheduler.py]: `num_update` useless here

File: python/mxnet/lr_scheduler.py [hunk 4]
```diff
@@ -136,3 +136,38 @@ def __call__(self, num_update):
...
+        super(PolyScheduler, self).__init__(base_lr)
+        assert isinstance(max_update, int)
+        if max_update < 1:
+            raise ValueError("maximum number of updates must be strictly positive")
+        self.base_lr_orig = self.base_lr
+        self.max_update = max_update
+        self.power = pwr
+        self.count = num_update
```
Comment (zhreshold) [inline python/mxnet/lr_scheduler.py]: same for self.count

File: python/mxnet/lr_scheduler.py [hunk 5]
```diff
@@ -136,3 +136,38 @@ def __call__(self, num_update):
...
+        self.base_lr = self.base_lr_orig
+
+    def __call__(self, num_update):
+        if num_update <= self.max_update:
+            self.base_lr = self.base_lr_orig * pow(1.0 - float(num_update) / float(self.max_update),
+                                                   self.power)
+        self.count += 1
```
Comment (zhreshold) [inline python/mxnet/lr_scheduler.py]: and here

Comment (ashokei) [reply python/mxnet/lr_scheduler.py]: thanks.. i see "count" is not being used, i removed it. Though, MultiFactorScheduler seems to be tracking count.

**Comment by zhreshold** [comment]
@szha test_io.test_LibSVMIter fails accasionally, we should host the data (http://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/multiclass/news20.t.bz2
) to avoid that. Add it to #9412 ??

**Comment by zhreshold** [comment]
@ashokei Try rebase and trigger the CI once more. We can merge once passed.

**Comment by ashokei** [comment]
@zhreshold all done, thanks!

File: python/mxnet/optimizer.py [hunk 6]
```diff
@@ -645,6 +645,195 @@ def update(self, index, weight, grad, state):
...
+@register
+class LBSGD(Optimizer):
+    """The Large Batch SGD optimizer with momentum and weight decay.
+
+    The optimizer updates the weight by::
+
+        state = momentum * state + lr * rescale_grad * clip(grad, clip_gradient) + wd * weight
+        weight = weight - state
+
+    For details of the update algorithm see :class:`~mxnet.ndarray.lbsgd_update` and
```
Comment (piiswrong) [inline python/mxnet/optimizer.py]: @ashokei @zhreshold 
Where is lbsgd_update defined? I don't see it.

Please add proper reference to relevant papar

File: python/mxnet/optimizer.py [hunk 7]
```diff
@@ -645,6 +645,195 @@ def update(self, index, weight, grad, state):
...
+        self.init_updates = begin_epoch * updates_per_epoch
+        self.num_epochs = num_epochs
+        # addl internal usage parameters and storage
+        self.lbmult = 1
+        self.cumgrads = {}
+        # for adaptive lr
+        self.adaptive = False
+        self.admult = 1  # adaptation constant
+
+    def create_state(self, index, weight):
```
Comment (piiswrong) [inline python/mxnet/optimizer.py]: Is this copied from SGD?
Why not inherit SGD instead?

File: python/mxnet/optimizer.py [hunk 8]
```diff
@@ -645,6 +645,195 @@ def update(self, index, weight, grad, state):
...
+    multi_precision: bool, optional
+       Flag to control the internal precision of the optimizer.
+       ``False`` results in using the same precision as the weights (default),
+       ``True`` makes internal 32-bit copy of the weights and applies gradients
+                in 32-bit precision even if actual weights used in the model have lower precision.`<
+                Turning this on can improve convergence and accuracy when training with float16.
+    warmup_strategy: string ('linear', 'power2', 'sqrt'. , 'lars'   default : 'linear')
+    warmup_epochs: unsigned, default: 5
+    batch_scale:   unsigned, default: 1 (same as batch size*numworkers)
+    updates_per_epoch: updates_per_epoch (default: 32, Default might not reflect true number batches per epoch. Used for warmup.)
```
Comment (piiswrong) [inline python/mxnet/optimizer.py]: Why use warmup_epochs and updates_per_epoch? Why not just warmup_updates?
Why should it have a default value?

File: python/mxnet/optimizer.py [hunk 9]
```diff
@@ -645,6 +645,195 @@ def update(self, index, weight, grad, state):
...
+    begin_epoch: unsigned, default 0, starting epoch.
```
Comment (piiswrong) [inline python/mxnet/optimizer.py]: What's starting epoch? What would it do before start epoch?

File: python/mxnet/optimizer.py (hunk 6 above)
Comment (ashokei) [reply python/mxnet/optimizer.py]: it is the update method in LBSGD class.  we can fix that to resolve to right method, the '_' is misleading.
the paper is here: 
https://arxiv.org/pdf/1708.03888.pdf

File: python/mxnet/optimizer.py (hunk 7 above)
Comment (zhreshold) [reply python/mxnet/optimizer.py]: @ashokei As suggested, could you change to inherit SGD and override `create_state_multi_precision`, `create_state`, `update`, `update_multi_precision` only if necessary.  Seems like you are mixing multi_precision part into the normal one.

File: python/mxnet/optimizer.py (hunk 8 above)
Comment (zhreshold) [reply python/mxnet/optimizer.py]: I guess it requires the epoch number to stop warming up, which does not depend on the number of updates.

File: python/mxnet/optimizer.py (hunk 9 above)
Comment (zhreshold) [reply python/mxnet/optimizer.py]: @ashokei please add more details describing the strategy.

Comment (chaseadams509) [reply python/mxnet/optimizer.py]: @piiswrong, The begin_epoch flag is because our data scientist saw that it was possible to stop a training partway, save it, then resume it again. We wanted to be able to have that option, which required passing in what the starting epoch is to do the learning rate decay calculations correctly.

**Comment by chenchao50** [comment]
Have you tested the accuracy of Resnet50(or AlexnetBN) in ImageNet using 'Lars' method ?

**Comment by rahul003** [comment]
Similar question as above. How do I use this? I'm unable to train resnet50 with lbsgd. What configuration works? I have an effective batch size of about 20k across 20 worker nodes

File: example/image-classification/common/fit.py [hunk 10]
```diff
@@ -180,40 +205,81 @@ def fit(args, network, data_loader, **kwargs):
...
+        args.monitor, pattern=".*") if args.monitor > 0 else None
 
-    if args.network == 'alexnet':
-        # AlexNet will not converge using Xavier
-        initializer = mx.init.Normal()
-    else:
-        initializer = mx.init.Xavier(
-            rnd_type='gaussian', factor_type="in", magnitude=2)
+    # A limited number of optimizers have a warmup period
+    has_warmup = {'lbsgd', 'lbnag'}
```
Comment (rahul003) [inline example/image-classification/common/fit.py]: It looks like LBNAG doesn't exist?

Comment (ashokei) [reply example/image-classification/common/fit.py]: @chaseadams509 can you please address above comments. thanks.

Comment (chaseadams509) [reply example/image-classification/common/fit.py]: Correct, lbnag was an optimizer we were experimenting with with incorporating the large-batch algorithm with Nesterov accelerated gradient. LBNAG was not showing the desired improvements, so we hadn't pushed it yet. Because of that, lbnag should probably be removed from this list.

Comment (rahul003) [reply example/image-classification/common/fit.py]: @chaseadams509 How large batch sizes did you experiment with?
